rclpy/src/rclpy/context.cpp

```diff
@@ -42,6 +42,7 @@ namespace rclpy
 {
 void shutdown_contexts()
 {
+  // graceful shutdown all contexts
   std::lock_guard guard{g_contexts_mutex};
   for (auto * c : g_contexts) {
     rcl_ret_t ret = rcl_shutdown(c);
```

Comment:
another way we can take is, we can remove the context from `g_contexts` once `rcl_shutdown` issued on it. so that `Context::shutdown` can check if it needs to issue `rcl_shutdown` on itself.

Reply:
That seems like a reasonable solution too, and matches `Context::shutdown()` removing the context from `g_contexts`.